lib/sycamore/sycamore/transforms/embed.py

```diff
@@ -19,44 +20,117 @@
     return document.text_representation if document.text_representation is not None else ""
 
 
-def _text_representation_is_empty(doc: Document) -> bool:
+def _text_representation_is_empty(doc: Union[Document, Element]) -> bool:
     return doc.text_representation is None or doc.text_representation.strip() == ""
 
 
+class EmbedderType(Enum):
+    SENTENCE_TRANSFORMER = "sentence_transformer"
+    OPENAI = "openai"
+    BEDROCK = "bedrock"
+
+
 class Embedder(ABC):
     def __init__(
         self,
         model_name: str,
+        embedder_type: EmbedderType,
         batch_size: Optional[int] = None,
-        model_batch_size: int = 100,
-        pre_process_document: Optional[Callable[[Document], str]] = None,
+        model_batch_size: Optional[int] = None,
+        pre_process_document: Optional[Callable[[Union[Document, Element]], str]] = None,
         device: Optional[str] = None,
+        client: Optional[OpenAIClient] = None,
     ):
         self.model_name = model_name
         self.batch_size = batch_size
-        self.model_batch_size = model_batch_size
         self.pre_process_document = pre_process_document if pre_process_document else _pre_process_document
-
         self.device = choose_device(device)
 
+        # Set default model batch sizes based on embedder type
+        if model_batch_size is None:
+            if embedder_type == EmbedderType.BEDROCK:
+                self.model_batch_size = 1  # Bedrock only supports single inputs
+            elif embedder_type == EmbedderType.OPENAI and isinstance(client, AzureOpenAIClient):
+                self.model_batch_size = 16  # For AzureOpenAIClient
+            else:
+                self.model_batch_size = 100  # Default for other embedders
+
+        else:
+            # Enforce maximum batch sizes for specific embedders
+            if embedder_type == EmbedderType.BEDROCK and model_batch_size > 1:
+                logger.warning(f"Reducing batch size from {model_batch_size} to 1 (Bedrock's maximum)")
+                self.model_batch_size = 1
+            elif (
+                embedder_type == EmbedderType.OPENAI and isinstance(client, AzureOpenAIClient) and model_batch_size > 16
+            ):
+                logger.warning(f"Reducing batch size from {model_batch_size} to 16 (AzureOpenAIClient's maximum)")
+                self.model_batch_size = 16
+            else:
+                self.model_batch_size = model_batch_size
+
     def __call__(self, doc_batch: list[Document]) -> list[Document]:
         return self.generate_embeddings(doc_batch)
 
-    @abstractmethod
     def generate_embeddings(self, doc_batch: list[Document]) -> list[Document]:
-        pass
+        """Handle batching and document processing logic in parent class"""
+
+        # Collect documents and elements to embed
+        text_to_embed = []
+        element_indices = defaultdict(set)
+        doc_indices = set()
+
+        # First pass: collect all texts that need embedding
```

Comment:
I think the following works, and it means that you don't have to worry about the algorithm for adding text to embed matches the algorithm for assigning the embeddings, e.g. you could swap the two if branches in the first loop and it would work fine.
```
        obj_for_embedding = []
        for doc in doc_batch:
            if not _text_representation_is_empty(doc):
                text_to_embed.append(self.pre_process_document(doc))
                obj_for_embedding.append(doc)

            if isinstance(doc, Document) and doc.get("elements"):
                for element in doc.elements:
                    if not _text_representation_is_empty(element):
                        text_to_embed.append(self.pre_process_document(element))
                        obj_for_embedding.append(elemnt)

         # Return early if nothing to embed
        if not text_to_embed:
            return doc_batch

        # Generate embeddings
        all_embeddings = []
        for text_batch in batched(text_to_embed, self.model_batch_size):
            batch_embeddings = self.embed_texts(text_batch)
            all_embeddings.extend(batch_embeddings)

        for i, embedding in enumerate(all_embeddings):
            obj_for_embedding[i].embedding = embedding
```